patches/nginx-1.11.2-resolve-hostsfile.patch

```diff
@@ -0,0 +1,373 @@
+# HG changeset patch
+# User Thibault Charbonnier <thibaultcha@me.com>
+# Date 1488252201 28800
+#      Mon Feb 27 19:23:21 2017 -0800
+# Branch resolve-hostsfile
+# Node ID e348d382e1ed78fba55261c1962625415ea21fd6
+# Parent  640f035293959b2d4b0ba5939d954bc517f57f77
+Resolver: parse hosts file entries
+
+The resolver directive can now take an optional 'hostsfile=<path>' option, such
+as:
+
+    resolver 8.8.4.4 hostsfile=/etc/hosts;
+
+Hosts parsed from the hosts file are considered valid forever. The behavior
+tries to be conservative, and only parses the hosts file when the option is
+provided, to enforce backwards compatibility.
+
+diff -r 640f03529395 -r e348d382e1ed src/core/ngx_resolver.c
+--- a/src/core/ngx_resolver.c	Fri Jan 27 19:06:35 2017 +0300
++++ b/src/core/ngx_resolver.c	Mon Feb 27 19:23:21 2017 -0800
+@@ -9,11 +9,12 @@
+ #include <ngx_core.h>
+ #include <ngx_event.h>
+ 
+-
+-#define NGX_RESOLVER_UDP_SIZE   4096
+-
+-#define NGX_RESOLVER_TCP_RSIZE  (2 + 65535)
+-#define NGX_RESOLVER_TCP_WSIZE  8192
++#define NGX_RESOLVER_HOSTSFILE_SIZE  4096
++
++#define NGX_RESOLVER_UDP_SIZE        4096
++
++#define NGX_RESOLVER_TCP_RSIZE       (2 + 65535)
++#define NGX_RESOLVER_TCP_WSIZE       8192
+ 
+ 
+ typedef struct {
+@@ -120,6 +121,8 @@
+     ngx_resolver_node_t *rn);
+ static void ngx_resolver_srv_names_handler(ngx_resolver_ctx_t *ctx);
+ static ngx_int_t ngx_resolver_cmp_srvs(const void *one, const void *two);
++static ngx_int_t ngx_resolver_parse_hostsfile(ngx_conf_t *cf,
++    ngx_resolver_t *r);
+ 
+ #if (NGX_HAVE_INET6)
+ static void ngx_resolver_rbtree_insert_addr6_value(ngx_rbtree_node_t *temp,
+@@ -128,7 +131,6 @@
+     struct in6_addr *addr, uint32_t hash);
+ #endif
+ 
+-
+ ngx_resolver_t *
+ ngx_resolver_create(ngx_conf_t *cf, ngx_str_t *names, ngx_uint_t n)
+ {
+@@ -243,6 +245,24 @@
+         }
+ #endif
+ 
++        if (ngx_strncmp(names[i].data, "hostsfile=", 10) == 0) {
++            r->hostsfile.len = names[i].len - 10;
++
++            if (r->hostsfile.len == 0) {
++                ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
++                                   "invalid parameter: %V", &names[i]);
++                return NULL;
++            }
++
++            r->hostsfile.data = ngx_resolver_dup(r, names[i].data + 10,
++                                                 r->hostsfile.len + 1);
```

Comment:
Hmm, looks like `ngx_conf_read_token()` actually [adds the `'\0'` for directive parameters](http://lxr.nginx.org/source/src/core/ngx_conf_file.c#0760). That is definitely not something I knew before! Closing this review.